### PR TITLE
Infimum application

### PR DESCRIPTION
**Pull request by rhysbalevicius**
#### Project Abstract

This application is in response to the [anti-collusion infrastructure RFP](https://github.com/w3f/Grants-Program/blob/master/docs/RFPs/Open/anti-collusion_infrastructure.md).

The goal of this proposal is to provide a minimum viable implementation of Vitalik Buterin’s proposed [“Minimal Anti-Collusion Infrastructure”](https://ethresear.ch/t/minimal-anti-collusion-infrastructure/5413) as a substrate pallet and CLI (for performing off-chain work, i.e. encrypting votes, processing messages, and generating proofs). 

The scope of this proposal is intended to be a proof of concept, of which contributes to the development of a genuine minimal anti-collusion infrastructure within the Substrate ecosystem. 

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (bank details via email _or_ BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (USDT) address in the application).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

___

**Additional comments**

I am aware that there is an existing open application in response to the RFP. I’m not sure what the proper etiquette is in this case, so please feel free to let me know if I should close this application in light of the existing one, and potentially re-open it in the future should it become appropriate. 

I am also aware that there have been two [previous](https://github.com/w3f/Grants-Program/pull/232) [applications](https://github.com/w3f/Grants-Program/pull/1144) for this RFP which had been accepted but then [terminated](https://github.com/w3f/Grants-Program/pull/645) due to [lack of progress](https://github.com/w3f/Grants-Program/pull/1830), and so it’s worth mentioning that the estimated timeline proposed in this application has been prudently influenced by that fact.

Thanks for your time and consideration.

### CODE REVIEW

**Comment by rhysbalevicius** [comment]
I have read and hereby sign the Contributor License Agreement.

**Review by Noc2 — DISMISSED** [review]
Thanks a lot for the application. This looks good to me, and I will share it with the rest of the team.

File: applications/infimum.md [hunk 1]
```diff
@@ -0,0 +1,228 @@
+# Infimum
+
+- **Team Name:** Apollos Collective
+- **Payment Address:** 0x9c10EbAEB989CFd239679d47B9100dc4ad57A536 (ERC20 USDC)
+- **[Level](https://github.com/w3f/Grants-Program/tree/master#level_slider-levels):** 2
+
+## Project Overview :page_facing_up:
+
+This application is in response to the [anti-collusion infrastructure RFP](https://github.com/w3f/Grants-Program/blob/master/rfps/open/anti-collusion_infrastructure.md).
```
Comment (semuelle) [inline applications/infimum.md:9]:
```suggestion
This application is in response to the [anti-collusion infrastructure RFP](https://github.com/w3f/Grants-Program/blob/master/docs/RFPs/Open/anti-collusion_infrastructure.md).
```

File: applications/infimum.md [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+| Number | Deliverable | Specification |
+| --- | --- | --- |
+| 0a. | License | MIT |
+| 0b. | Documentation | Inline documentation. Basic guide explaining how to interact with the pallet will be provided in the README.  |
+| 0c. | Testing | Unit tests, GitHub actions CI workflow, and brief guide for running tests locally |
+| 0d. | Docker | Dockerfiles and docker-compose.yml for running a development environment which locally spins up a node and frontend template for observing events, calling pallet extrinsics, and performing state queries. |
+| 1. | Substrate pallet | Methods 1.ii.a to 1.ii.i (listed under deliverables in the project overview) without verification functionality provided by Groth16 proving system.  |
+
+### Milestone 2 Example — On-chain verification logic and circuits
```
Comment (semuelle) [inline applications/infimum.md:173]:
```suggestion
### Milestone 2 — On-chain verification logic and circuits
```

**Review by semuelle — CHANGES_REQUESTED** [review]

**Review by Noc2 — APPROVED** [review]

**Comment by rhysbalevicius** [comment]
Hi @semuelle, thanks for the helpful suggested changes. I've gotten those changes in, as well as a few other minor corrections

**Review by semuelle — CHANGES_REQUESTED** [review]
Thank you for the application, @rhysbalevicius. I am generally happy to support it, but could you add an example or tutorial that lets users quickly and easily test the architecture, as suggested by the RFP? Also, could you expand a bit on the contraints and limitations you mentioned, e.g. to what extent they are _due to_ the proposed architecture and to what extent they can be improved upon given this architecture? I'd be particularly interested in weighted votes.

> I am aware that there is an existing open application in response to the RFP. I’m not sure what the proper etiquette is in this case, so please feel free to let me know if I should close this application in light of the existing one, and potentially re-open it in the future should it become appropriate.

We are always happy to support more than one application/grant per RFP, because, as you can see, not all grants end in a successful delivery.

**Comment by rhysbalevicius** [comment]
Hi @semuelle, thank you for your feedback.

> could you add an example or tutorial that lets users quickly and easily test the architecture, as suggested by the RFP

If I understand your question correctly, milestones 1.0d, 1.0b, 2.0b, 3.0b, and in particular 3.2 (“all necessary setup and usage instructions into a readable and user friendly docsite”) encompass exactly what you’re asking for. That being said, I’d be happy to amend the proposal to specifically mention tutorial as a deliverable if you think that would be more appropriate. 

> could you expand a bit on the contraints and limitations you mentioned, e.g. to what extent they are due to the proposed architecture and to what extent they can be improved upon given this architecture

The intent here is to minimize the level of complexity introduced into the scope of deliverables. By limiting complexity the hope is to help to limit the potential surface area for bugs, provide more of a modular framework to build off of, and thereby provide a solid foundation for future work. 

The specific limitation you singled out is not a consequence of the architecture itself, but primarily a deliberate choice to reduce the complexity of the scope of deliverables. In order to permit weighted votes one would need to modify the circuits (delivered in milestone 2.2) used. Support for weighted voting, etc, is proposed under the future work section (2.iii). Ideally, in a “production ready” future v1 incarnation of the system, a coordinator would be able to easily create different types of polls which have weighted, ranked choice, etc ballots. 

On the other hand, 
- Processing in batches is indeed a consequence of the architecture. In particular, inputs to a circuit must be known at compile time (yet the number of participants may vary from poll to poll). This constraint could potentially be addressed by leveraging “recursive SNARKs” (see future work section 2.ii), or in theory an alternative architecture altogether (see future work section 1.i). 

- Use of a trusted setup is a consequence of using Groth16, which is non-transparent. Naturally, improving upon this would rely on introducing a transparent system (this is noted under future work section 2.i).

- The choice for an individual coordinator to have at most one active poll at a time is a consequence of the storage layout (a coordinator may only have a single active set of keys at a time). 

The potential solution(s) for each of these has been intentionally excluded specifically for the reasons mentioned above.

**Comment by nikw3f** [comment]
Thanks for the application. Can you provide more details on how the CLI tool will interface with the Substrate pallet?

**Comment by rhysbalevicius** [comment]
Hi @nikw3f, thanks for taking the time to review. My apologies for the delayed response.

> Can you provide more details on how the CLI tool will interface with the Substrate pallet?

The CLI will mainly rely on [polkadot-api](https://github.com/polkadot-js/api) as a dependency in order to call extrinsics and query state where necessary. 

Within the context of the long term vision for the project we do not see the CLI as the primary way to interact with the system, but the shortest path to providing developers and reviewers alike a way to test the system, as well as dissect the inner workings and flow of interactions. Therefore we plan on designing the CLI to operate as a thin wrapper which consumes a module of functions. 

In future work we would like to see this module consumed from within the context of a dApp as we believe this would provide an enriched user experience and lower the barrier to entry in terms of permitting interactions with the system. 

Let me know if there are any details in particular you are wondering about.

**Comment by keeganquigley** [comment]
Thanks for the application @rhysbalevicius and for adding all the details about the pallet's public methods. I'd also be generally happy to support it, but I have one question: Does this mean that weighted voting won't be in the current scope since users can only cast a vote of weight 1?

**Comment by rhysbalevicius** [comment]
Hi @keeganquigley, thanks for taking the time to review. 

> Does this mean that weighted voting won't be in the current scope since users can only cast a vote of weight 1?

That’s correct. While the proposed design does not limit this as being added as a feature (and indeed is listed on the roadmap for future work), this was a deliberate choice to limit complexity in the initial version. 

To elaborate a bit further on the thought process which went into this decision, the problem I’m trying to avoid is the following: how should “voting power” be decided? 

For example, during the sign up phase voters could use a special voucher which encodes their voting power. Alternatively, a coordinator could be responsible for assigning the voting power to individual voters after signup. However, both of these approaches add friction to the overall user experience. Moreover, they are opinionated in how external systems may interface with them, and so in particular I’d like to be careful here in order to avoid limiting potential integrations. 

The correct design for a flexible but sufficiently powerful interface does not seem immediately obvious to me and requires additional deliberation (ideally with feedback from the community). Since this feature could instead be built on top of a system which provides uniform voting power this seemed like the clear choice in order to move forward. 

That all being said, I can definitely appreciate why this feature would be important to the community, and since it’s been brought up a second time now I’m wondering if excluding it would be considered a dealbreaker. I’m definitely open to suggestion for potential amendments to the proposal which address the above concerns!

**Review by keeganquigley — APPROVED** [review]
Thanks for the clarification @rhysbalevicius I'm willing to approve it as is and give it a shot.

**Review by semuelle — APPROVED** [review]
Thank you for the updates and the responses, @rhysbalevicius. Happy to proceed.

**Comment by rhysbalevicius** [comment]
> Thank you for the updates and the responses, @rhysbalevicius. Happy to proceed.

Wonderful. Thank you all for the support 🎉

**Comment by keeganquigley** [comment]
Hi @rhysbalevicius how is milestone 1 coming along?

**Comment by rhysbalevicius** [comment]
Hey @keeganquigley. M1 is coming along well, just doing a bit of final clean up before submitting for review. You should expect to see a PR from me this week!

**Comment by keeganquigley** [comment]
Hi @rhysbalevicius are you making progress on milestone 2?

**Comment by rhysbalevicius** [comment]
> Hi @rhysbalevicius are you making progress on milestone 2?

Hey @keeganquigley,

Progress is being made on M2, albeit at a slower pace than anticipated. There are some ongoing personal circumstances that have unfortunately reduced the frequency at which I am able to contribute to the project.

I anticipate submitting M2 for review sometime early to mid-next quarter.

**Comment by PieWol** [comment]
Hey @rhysbalevicius ,
it's been a while since we heard from you. After checking the latest amendment to the timelines it looks like the second milestone is already a few months overdue. Do you have any updates to share with us by now? 
In case you chose to work on different projects and lost track of this grant then we will go ahead and terminate it.

Looking forward to hearing from you.

**Comment by rhysbalevicius** [comment]
Hi @PieWol, I greatly appreciate the ongoing patience with the timeline for this project.

In retrospect the amended timeline proposed was ambiguous but was intended to extend until this quarter, my apology for the lack of clarification.

At any rate, I am in the process of wrapping up the current milestone and will be submitting it for evaluation during the first week of December. 

Thanks again,
Rhys

**Pull request merged**